File: scripts/import-residents.mjs
#!/usr/bin/env node
/**
 * CLI 工具：直接從 JSON 檔案匯入住戶資料到資料庫
 * 用法：node scripts/import-residents.mjs <json_file_path>
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { drizzle } from 'drizzle-orm/mysql2';
import { residents } from '../drizzle/schema.js';
import mysql from 'mysql2/promise';
import { createConnection } from 'mysql2/promise';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// 從環境變數讀取資料庫連接字串
const DATABASE_URL = process.env.DATABASE_URL;

if (!DATABASE_URL) {
  console.error('❌ 錯誤：DATABASE_URL 環境變數未設定');
  process.exit(1);
}

// 解析 MySQL 連接字串
const parseConnectionString = (url) => {
  // mysql://user:password@host:port/database
  const match = url.match(/mysql:\/\/([^:]+):([^@]+)@([^:]+):(\d+)\/(.+)/);
  if (!match) {
    throw new Error('無效的 DATABASE_URL 格式');
  }
  return {
    user: match[1],
    password: match[2],
    host: match[3],
    port: parseInt(match[4]),
    database: match[5],
  };
};

const importResidents = async (filePath) => {
  try {
    // 讀取 JSON 檔案
    if (!fs.existsSync(filePath)) {
      console.error(`❌ 檔案不存在：${filePath}`);
      process.exit(1);
    }

    const fileContent = fs.readFileSync(filePath, 'utf-8');
    const data = JSON.parse(fileContent);

    if (!Array.isArray(data)) {
      console.error('❌ JSON 檔案必須是陣列格式');
      process.exit(1);
    }

    console.log(`📁 讀取檔案：${filePath}`);
    console.log(`📊 準備匯入 ${data.length} 筆資料...\n`);

    // 連接資料庫
    const config = parseConnectionString(DATABASE_URL);
    const connection = await createConnection(config);
    const db = drizzle(connection);

    let successCount = 0;
    let errorCount = 0;
    const errors = [];

    // 逐筆匯入
    for (let i = 0; i < data.length; i++) {
      const resident = data[i];
      try {
        // 驗證必填欄位
        if (!resident.unitNumber) {
          throw new Error('戶號為必填欄位');
        }
        if (!resident.ownerName) {
          throw new Error('區權人姓名為必填欄位');
        }

        // 檢查戶號是否重複
        const existing = await db
          .select()
          .from(residents)
          .where((r) => r.unitNumber === resident.unitNumber)
          .limit(1);

        if (existing.length > 0) {
          throw new Error(`戶號 ${resident.unitNumber} 已存在`);
        }

        // 插入資料
        await db.insert(residents).values({
          unitNumber: resident.unitNumber,
          ownerName: resident.ownerName,
          ownerPhone: resident.ownerPhone || null,
          coResident1Name: resident.coResident1Name || null,
          coResident1Phone: resident.coResident1Phone || null,
          coResident2Name: resident.coResident2Name || null,
          coResident2Phone: resident.coResident2Phone || null,
          coResident3Name: resident.coResident3Name || null,
          coResident3Phone: resident.coResident3Phone || null,
          coResident4Name: resident.coResident4Name || null,
          coResident4Phone: resident.coResident4Phone || null,
          carParkingNumber: resident.carParkingNumber || null,
          carPlateNumber: resident.carPlateNumber || null,
          motorcycleParkingNumber: resident.motorcycleParkingNumber || null,
          motorcyclePlateNumber: resident.motorcyclePlateNumber || null,
          bicycleParkingNumber: resident.bicycleParkingNumber || null,
          address: resident.address || null,
          emergencyContactName: resident.emergencyContactName || null,
          emergencyContactPhone: resident.emergencyContactPhone || null,
          emergencyContactRelation: resident.emergencyContactRelation || null,
          emergencyContactAddress: resident.emergencyContactAddress || null,
          emergencyContact2Name: resident.emergencyContact2Name || null,
          emergencyContact2Phone: resident.emergencyContact2Phone || null,
          emergencyContact2Relation: resident.emergencyContact2Relation || null,
          emergencyContact2Address: resident.emergencyContact2Address || null,
          notes: resident.notes || null,
        });

        successCount++;
        console.log(`✓ [${i + 1}/${data.length}] ${resident.unitNumber} - ${resident.ownerName}`);
      } catch (err) {
        errorCount++;
        const errorMsg = err instanceof Error ? err.message : String(err);
        errors.push({
          index: i,
          unitNumber: resident.unitNumber,
          ownerName: resident.ownerName,
          error: errorMsg,
        });
        console.log(`✗ [${i + 1}/${data.length}] ${resident.unitNumber} - ${errorMsg}`);
      }
    }

    // 關閉連接
    await connection.end();

    // 輸出統計結果
    console.log(`\n${'='.repeat(60)}`);
    console.log(`匯入完成！`);
    console.log(`✓ 成功：${successCount} 筆`);
    console.log(`✗ 失敗：${errorCount} 筆`);
    console.log(`${'='.repeat(60)}`);

    if (errors.length > 0) {
      console.log(`\n❌ 失敗詳情：`);
      errors.forEach((err) => {
        console.log(`  - ${err.unitNumber} (${err.ownerName})：${err.error}`);
      });
    }

    process.exit(errorCount === 0 ? 0 : 1);
  } catch (err) {
    console.error('❌ 匯入失敗：', err instanceof Error ? err.message : String(err));
    process.exit(1);
  }
};

// 取得命令列參數
const args = process.argv.slice(2);
if (args.length === 0) {
  console.log('用法：node scripts/import-residents.mjs <json_file_path>');
  console.log('範例：node scripts/import-residents.mjs /home/ubuntu/upload/B棟_converted.json');
  process.exit(1);
}

const filePath = args[0];
importResidents(filePath);

File: import_residents.py
import csv
import psycopg2
import json
from datetime import datetime

conn_str = "postgresql://neondb_owner:npg_knpNu8lB2FWR@ep-dry-glade-azg6pbwu-pooler.c-3.ap-southeast-1.aws.neon.tech/neondb?sslmode=require"

print("Connecting to database...")
conn = psycopg2.connect(conn_str)
conn.autocommit = False
cur = conn.cursor()

print("Reading CSV file...")
with open('/home/ubuntu/upload/residents_2026-07-21(2).csv', 'r', encoding='utf-8') as f:
    reader = csv.reader(f, delimiter='\t')
    rows = list(reader)

headers = rows[0]
data_rows = rows[1:]
print(f"Total data rows: {len(data_rows)}")

# Map headers to column indices
header_map = {h: i for i, h in enumerate(headers)}

# Column mapping: CSV header -> DB column (same names)
db_columns = [
    'unitNumber', 'ownerName', 'ownerPhone', 'address',
    'coResident1Name', 'coResident1Phone', 'coResident2Name', 'coResident2Phone',
    'coResident3Name', 'coResident3Phone', 'coResident4Name', 'coResident4Phone',
    'carParkingNumber', 'carPlateNumber', 'motorcycleParkingNumber', 'motorcyclePlateNumber',
    'bicycleParkingNumber', 'squareMeters', 'waterMeterNumber', 'electricityMeterNumber',
    'moveInDate', 'emergencyContactName', 'emergencyContactPhone', 'emergencyContactRelation',
    'emergencyContact2Name', 'emergencyContact2Phone', 'emergencyContact2Relation',
    'emergencyContact2Address', 'notes'
]

success_count = 0
error_count = 0
errors = []

print("Importing residents...")
for i, row in enumerate(data_rows):
    try:
        # Extract values using header map
        values = {}
        for col in db_columns:
            idx = header_map.get(col)
            if idx is not None and idx < len(row):
                val = row[idx].strip() if row[idx] else None
                values[col] = val if val else None
        
        # Handle moveInDate - convert to date or null
        move_in_date = values.get('moveInDate')
        if move_in_date:
            # Try to parse as date
            try:
                datetime.strptime(move_in_date, '%Y-%m-%d')
            except ValueError:
                move_in_date = None
        values['moveInDate'] = move_in_date
        
        # Skip rows with empty unitNumber
        if not values.get('unitNumber'):
            print(f"  Skipping row {i+1}: empty unitNumber")
            continue
        
        # Check if unitNumber already exists
        cur.execute("SELECT id FROM residents WHERE \"unitNumber\" = %s", (values['unitNumber'],))
        existing = cur.fetchone()
        if existing:
            print(f"  Skipping row {i+1}: unitNumber {values['unitNumber']} already exists (id={existing[0]})")
            continue
        
        # Insert into residents table
        col_names = ', '.join([f'"{c}"' for c in db_columns])
        placeholders = ', '.join(['%s'] * len(db_columns))
        sql = f'INSERT INTO residents ({col_names}) VALUES ({placeholders})'
        vals = [values.get(c) for c in db_columns]
        
        cur.execute(sql, vals)
        
        success_count += 1
        if (success_count + error_count) % 50 == 0:
            print(f"  Progress: {success_count + error_count}/{len(data_rows)} (success={success_count}, error={error_count})")
            
    except Exception as e:
        error_count += 1
        unit = row[header_map.get('unitNumber', 1)] if len(row) > 1 else f"row {i+1}"
        errors.append(f"Row {i+1} ({unit}): {str(e)[:200]}")
        print(f"  ERROR row {i+1} ({unit}): {str(e)[:100]}")

conn.commit()
print(f"\n=== Import Summary ===")
print(f"Total rows: {len(data_rows)}")
print(f"Successfully imported: {success_count}")
print(f"Skipped/Errors: {error_count}")

if errors:
    print(f"\nFirst 10 errors:")
    for e in errors[:10]:
        print(f"  {e}")

# Verify
cur.execute("SELECT COUNT(*) FROM residents")
count = cur.fetchone()[0]
print(f"\nTotal residents in database: {count}")

cur.close()
conn.close()
print("\nDone!")
